Automattic/woocommerce-payments
# WCPay base API Request class

Fixes #5056.

#### Changes proposed in this Pull Request

The purpose of this PR is to add the foundation class (and all methods) where you could simply execute WCPay API endpoint   using the class like a builder where you could chain function calls and where you could build fully working API requests to WCPay API server. The approach that will be used is combination of DTOs and Value Objects where data could be setter in a way that will provide argument validation. 

This PR will add all necessary classed and functionality where you could create request and all necessary checks will be performed. The main purpose is to create a functionality that will look something like this:

```
$create_charge = new Create_Charge();
		$create_charge
			->set_amount( 100 )
			->set_source_id( 'source_id' );
$result = $this->payments_api_client->send_wcpay_request( $create_charge );
```
		
With this approach, the type-safety will be ensured and building API requests will be way easier. 

Instead of relaying on arrays when response is returned from server, a new DTO have been added that extends request and extend it by functions that could be used somewhere else. The main purpose of this DTO is that we could extend request class and include DTOs for API responses, where we will ensure safety of the parameters that we got from the API response.

This PR is only one PR where others will be created once we conclude on how request class should look like. To have a better understanding, take a look at Create_Charge class where basic request have been created.

Another thing that is added are interfaces. The main purpose of them is to create a set of rules that we could use if we decide to change the request (or response) class sometime in the future. There are abstract or base classed that defines core functionality that will be used from all requests, DTOs, and value objects. For every request class, there is a generic `__set` method that is inherited from the base class to every class that extends it. the `__set` function will try to find an internal `set_SOMETHING` function that will hold validation data (if any) and if a method doesn't exist, the normal way of setting a variable will be used.

This PR has been edited, and some code have been removed. The reason behind this is that there is too much code that solves too many problems, and separating or creating multiple PRs will be easier to track and work on.







#### Testing instructions

1. Create new create charge API request using added class ( `new Create_Charge()` )
2. Make sure that exception will be thrown when charge amount is not number
3. Make sure that exception will be thrown when amount is below zero
4. Make sure that exception will be thrown if source id is empty
5. Fill class with correct data
6. Execute WCPay request by calling `$this->payments_api_client->send_wcpay_request()` and pass created class from step 1
7. Make sure that return of this function is Response DTO that was added in this PR

-------------------


**Post merge**
